tlog/tlogserver/server/server.go

```
@@ -97,23 +97,44 @@ func (s *Server) handle(conn net.Conn) error {
 			return err
 		}
 
+		// vdiskID
 		vdiskID, err := tlb.VdiskID()
 		if err != nil {
 			log.Debugf("failed to get vdisk ID: %v", err)
 			return err
 		}
 
+		if tlc == nil {
```

I think we are having a miss communication here.

I'm talking about the same connection. The Vdisk isn't bound to the tcp connection. Reading from the code, we are getting the vdiskID from the received value, each for loop iteration, meaning that there is no guarantee (as far that I know), that even within the same connection, we are sending always the same vdiskID.

Perhaps you never expect that a single connection returns different vdiskID, but it is certainly not validated here. And you know... always validate user input!

So either:

+ We need to make this code work with different vdisks on the same connection;
+ You need to validate that the vdiskID is always the same, and block the transaction in case it is meant for a different vdisk;

But the code as it right now is, would just silently accept this, and push the blocks to the wrong channel, as stated in my original comment...